overlord/devicestate/devicestate.go

```diff
@@ -637,3 +637,48 @@ func Remodeling(st *state.State) bool {
 	}
 	return false
 }
+
+type recoverySystemSetup struct {
+	// Label of the recovery system
+	Label string `json:"label"`
+	// Directory inside the seed filesystem where the recovery system files
+	// are kept
+	Directory string `json:"directory"`
+	// NewCommonFiles is a list of new files added to the shared snaps
+	// directory in seed
+	NewCommonFiles []string `json:"new-common-files"`
+	// SnapSetupTasks is a list of tasks that carry snap setup information,
+	// relevant only during remodel
+	SnapSetupTasks []string `json:"snap-setup-tasks"`
+}
+
+func createRecoverySystemTasks(st *state.State, label string, snapSetupTasks []string) *state.TaskSet {
+	create := st.NewTask("create-recovery-system", fmt.Sprintf("Create recovery system with label %q", label))
+	// the label we want
+	create.Set("recovery-system-setup", &recoverySystemSetup{
+		Label: label,
+		// IDs of the tasks carrying snap-setup
+		SnapSetupTasks: snapSetupTasks,
+	})
+
+	finalize := st.NewTask("finalize-recovery-system", fmt.Sprintf("Finalize recovery system with label %q", label))
+	finalize.WaitFor(create)
+	// finalize needs to know the label too
+	finalize.Set("recovery-system-setup-task", create.ID())
+	return state.NewTaskSet(create, finalize)
+}
+
+func CreateRecoverySystem(st *state.State, label string) (*state.Change, error) {
```

Comment:
it's a bit atypical to create changes in the mangers themselves, do we have other examples?

Reply:
Hm Remodel() creates a change, and pretty sure I saw some other code in snapstate that does that too.

Reply:
FWIW, this is really only needed to be called from daemon in #10255 at this point.